test/issue-189-default-values.cpp

```diff
@@ -0,0 +1,49 @@
+#include <iostream>
+#include <nlohmann/json-schema.hpp>
+
+using nlohmann::json;
+using nlohmann::json_schema::json_validator;
+
+static const json person_schema = R"(
+{
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "title": "A rectangle",
+    "properties": {
+        "width": {
+            "$ref": "#/definitions/length",
+			"default": 20
+        },
+        "height": {
+            "$ref": "#/definitions/length"
+        }
+    },
+	"definitions": {
+        "length": {
+			"type": "integer",
+			"default": 10
+    	}
+	},
+    "additionalProperties": false,
+    "type": "object"
+})"_json;
+
+int main(void)
+{
+	json_validator validator{};
+	validator.set_root_schema(person_schema);
+
+	{
+		json empty_rectangle = R"({})"_json;
+
+		const auto default_patch = validator.validate(empty_rectangle);
+		const auto actual = empty_rectangle.patch(default_patch);
+
+		// height must be 10 according to the default specified in the length definition while width must be 10 overridden by the width element
+		const auto expected = R"({"height":10,"width":20})"_json;
+		if (actual != expected) {
+			std::cerr << "Patch with defaults contains wrong value: '" << default_patch << "' instead of expected '" << expected.dump() << "'" << std::endl;
+			return 1;
+		}
+	}
+	return 0;
+}
```

Comment:
Could you add a test where two stacked/recursive `$refs` are used? And see whether it still works. I'm afraid that there is some late evaluation of not-yet loaded references which will break this mechanism.

Reply:
Can you provide an example with stacked/recursive $refs and default values? I am not able to figure out a useful one...

Reply:
In addition to that, it should use other files to see whether late evaluation works. (You might laugh, but I don't remember whether I do late evaluation of references or not...) Let me check your branch and see what I can do.

Reply:
I adapted your example, it doesn't work if there is one more indirection. Could you check whether you can make it work. https://github.com/pboettch/json-schema-validator/tree/martin-koch/master

Reply:
OK, I'm looking at it now, I think there is no need to copy the schema. We just need to get the earliest default value. I'll try.

Reply:
Got it. I'll push on your master branch via this pull-request. Please test and review. And you're right, we need to copy the ref-schema when a (new) default is present.

Reply:
Damn it. There is a corner case regarding the copy of the ref... When the same ref on a ref is used, the default value won't be used... hmm, I don't understand myself, let's think about it a bit more.

Reply:
No, it's OK. The copy is actually what's need to be done. Thanks. If you're OK with these changes, could you update the README with this use-case and than I will happily merge it.

Reply:
I updated the readme and fixed an issue with the simpler example (segmentation fault) which occurred after change 40f5d29 since get_or_create_ref may return an instance of schema_ref **or** an instance of type_schema.